### PR TITLE
Implement API priorities (reading)

### PR DESCRIPTION
**Pull request by Azure7111**
https://community.openproject.org/work_packages/17411

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://feature-api-read-priorities.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://feature-api-read-priorities.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by myabc** [comment]
@Azure7111 `rubocop -a` may help quieten Hound :smile_cat:

**Comment by NobodysNightmare** [comment]
@myabc:

```
$ rubocop -a
...
1436 files inspected, 20751 offenses detected, 16793 offenses corrected
$ git status | grep modified: | wc -l
350
```

:joy_cat: 

But we are on it to fix some additional styling errors...

P.S.: What we really did was

```
$ git diff --files-only our_branch base | grep .rb | xargs rubocop
```

**Comment by myabc** [comment]
@NobodysNightmare very nice!

**Comment by myabc** [comment]
but I think you mean:

```
git diff --files-only our_branch base | grep .rb | xargs rubocop -a
```

**Comment by NobodysNightmare** [comment]
Nope, we went the extra mile to understand what we are doing ^^

File: spec/lib/api/v3/priorities/priority_representer_spec.rb [hunk 1]
```diff
@@ -32,19 +32,43 @@
...
+      end
+      it 'should have a position' do
+        should have_json_path('position')
+      end
+      it 'should have a default flag' do
+        should have_json_path('isDefault')
+      end
+      it 'should have an active flag' do
+        should have_json_path('isActive')
+      end
```
Comment (ulferts) [inline spec/lib/api/v3/priorities/priority_representer_spec.rb:71]: @Azure7111, @NobodysNightmare you could (and should) tests for the specific values. e.g. 

```
it 'should have an id' do
  is_expected.to be_json_eql(priority.id.to_json).at_path('id')
end
```

Comment (myabc) [inline spec/lib/api/v3/priorities/priority_representer_spec.rb:71]: (`is_expected.to` does the same thing as `should`, but probably reads slightly better)

File: lib/api/v3/work_packages/work_package_representer.rb [hunk 2]
```diff
@@ -231,9 +239,16 @@ class WorkPackageRepresenter < ::API::Decorators::Single
                  },
                  setter: -> (value, *) { represented.description = value['raw'] },
                  render_nil: true
-        property :priority, getter: -> (*) { priority.try(:name) }, render_nil: true
```
Comment (ulferts) [inline lib/api/v3/work_packages/work_package_representer.rb:234]: @0xF013, @NobodysNightmare  this is going to destroy displaying the priority in the WP pane.

File: lib/api/v3/utilities/path_helper.rb [hunk 3]
```diff
@@ -70,6 +70,10 @@ def self.priorities
             "#{root}/priorities"
           end
 
+          def self.priority(id)
+            "#{priorities}/#{id}"
```
Comment (ulferts) [inline lib/api/v3/utilities/path_helper.rb:74]: This is untested code :scream:

-> spec/lib/api/v3/utilities/path_helper_spec.rb

File: spec/requests/api/v3/priority_resource_spec.rb [hunk 4]
```diff
@@ -28,20 +28,22 @@
 
 require 'spec_helper'
 require 'rack/test'
+require 'pry'
```
Comment (ulferts) [inline spec/requests/api/v3/priority_resource_spec.rb]: @Azure7111, @NobodysNightmare 

No, it's not required.

File: spec/requests/api/v3/priority_resource_spec.rb [hunk 5]
```diff
@@ -55,5 +57,61 @@
...
+    subject(:response) { last_response }
+
+    let(:get_path) { "/api/v3/priorities/#{priorities.first.id}" }
+
+    context 'logged in user' do
+      before do
+        allow(User).to receive(:current).and_return current_user
+        member = FactoryGirl.build(:member, user: current_user, project: project)
+        member.role_ids = [role.id]
+        member.save!
```
Comment (ulferts) [inline spec/requests/api/v3/priority_resource_spec.rb]: @Azure7111 @NobodysNightmare 

Not an issue, just a pointer -> there is a more condensed way of writing this: https://github.com/opf/openproject/blob/dev/spec/models/query/results_spec.rb#L71

**Comment by ulferts** [comment]
@Azure7111 @NobodysNightmare generally looking good.

Thanks for rubocopifying :balloon:

Please let me know who is going to address the details pane incompatibility issue and in which PR.

File: lib/api/v3/utilities/path_helper.rb (hunk 3 above)
Comment (NobodysNightmare) [inline lib/api/v3/utilities/path_helper.rb:74]: not anymore ;)

**Comment by ulferts** [comment]
One other thing I noticed when taking a look at the read part (a little late, I know) is that the priority is not embedded. That is inconsistent.

**Comment by NobodysNightmare** [comment]
@ulferts @0xF013 I tried to look into the JS code in the hope to quickly identify where I needed to replace the access to `priority` with the access to `_links/priority/title`... however, I had no luck...

@0xF013 Would you mind fixing the WP pane? Do you need a work package for that? Would you do it in an extra PR?

**Comment by 0xF013** [comment]
@NobodysNightmare I can do it. Should I hijack this branch?

**Comment by NobodysNightmare** [comment]
You are welcome to do that :)

I will just add the missing embedding that @ulferts mentioned.

edit: I am just done with that. The branch is all yours now ;-)

**Comment by ulferts** [comment]
Thanks everybody. Looks good.

**Pull request merged**